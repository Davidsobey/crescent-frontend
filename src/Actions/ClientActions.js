import ClientConstants from '../Constants/ClientConstants';
import ClientService from '../Services/ClientService';
import AlertActions from './AlertActions';
import history from '../Helpers/History';

function create(client) {
  function request() {
    return { type: ClientConstants.CREATE_REQUEST, client };
  }
  function success() {
    return { type: ClientConstants.CREATE_SUCCESS, client };
  }
  function failure(error) {
    return { type: ClientConstants.CREATE_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request({ client }));
    ClientService.create(client).then(
      () => {
        dispatch(success(client));
        history.push('/client/list');
        dispatch(AlertActions.success('Client created.'));
      },
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function subscribe(subscription) {
  function request() {
    return { type: ClientConstants.SUBSCRIBE_REQUEST, subscription };
  }
  function success() {
    return { type: ClientConstants.SUBSCRIBE_SUCCESS, subscription };
  }
  function failure(error) {
    return { type: ClientConstants.SUBSCRIBE_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request({ subscription }));
    ClientService.subscribe(subscription).then(
      () => {
        dispatch(success(subscription));
        history.push(`/client/${subscription.clientId}.list`);
        dispatch(AlertActions.success('Subscription created.'));
      },
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function getById(id) {
  function request() {
    return { type: ClientConstants.GETBYID_REQUEST };
  }
  function success(client) {
    return { type: ClientConstants.GETBYID_SUCCESS, client };
  }
  function failure(error) {
    return { type: ClientConstants.GETBYID_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request());

    ClientService.getById(id).then(
      client => dispatch(success(client)),
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function getAll() {
  function request() {
    return { type: ClientConstants.GETALL_REQUEST };
  }
  function success(clients) {
    return { type: ClientConstants.GETALL_SUCCESS, clients };
  }
  function failure(error) {
    return { type: ClientConstants.GETALL_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request());

    ClientService.getAll().then(
      clients => dispatch(success(clients)),
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function deleteClient(id) {
  function request() {
    return { type: ClientConstants.DELETE_CLIENT_REQUEST, id };
  }
  function success() {
    return { type: ClientConstants.DELETE_CLIENT_SUCCESS, id };
  }
  function failure(error) {
    return { type: ClientConstants.DELETE_CLIENT_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request());

    ClientService.deleteClient(id).then(
      () => {
        dispatch(success(id));
        dispatch(AlertActions.success('Client deleted.'));
      },
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function loadClient(id) {
  function request() {
    return { type: ClientConstants.LOAD_CLIENT_REQUEST };
  }
  function success(client) {
    return { type: ClientConstants.LOAD_CLIENT_SUCCESS, client };
  }
  function failure(error) {
    return { type: ClientConstants.LOAD_CLIENT_FAILURE, error };
  }

  return (dispatch) => {
    dispatch(request(id));

    ClientService.getClient(id).then(
      (client) => {
        dispatch(success(client));
      },
      (error) => {
        dispatch(failure(error));
        dispatch(AlertActions.error(error));
      },
    );
  };
}

function editClient(values) {
  return (dispatch) => {
    ClientService.editClient(values).then(
      () => {
        history.push('/client/list');
        dispatch(AlertActions.success(`Client ${values.name} edited.`));
      },
      (error) => {
        dispatch(AlertActions.error(error));
      },
    );
  };
}

const clientActions = {
  create,
  getAll,
  getById,
  subscribe,
  deleteClient,
  loadClient,
  editClient,
};

export default clientActions;
