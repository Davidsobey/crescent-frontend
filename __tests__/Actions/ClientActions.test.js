import configureStore from 'redux-mock-store';

// Actions to be tested
import ClientActions from '../../src/Actions/ClientActions';

const mockStore = configureStore();
const store = mockStore();

describe('ClientActions', () => {
  beforeEach(() => { // Runs before each test in the suite
    store.clearActions();
  });

  test('Dispatches the correct action and payload', () => {
    expect(store.getActions()).toMatchSnapshot();
  });

});