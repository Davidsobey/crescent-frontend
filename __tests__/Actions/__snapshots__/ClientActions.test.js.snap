// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientActions Dispatches the correct action and payload 1`] = `Array []`;
